Optimize ddoc cache opener

eiri opened the pull request:
This PR changes how the cache handles opening and expiration of the ddocs.

Now each open_doc request spawns a separated `gen_server` process
under control of a custom supervisor. This process responds
to the requestors itself, so the response time do not depend
on how busy ddoc_opener's message queue is.

Opener process for none-revisioned ddocs made long-living, it does
a periodic check for a newest revision of its ddoc and in case
of an updated revision fetches the ddoc, keeping the cache warm.

BugsID: 28132

eiri committed:
Move ddoc open functionality into separate process

Each open_doc request now opens a separated `gen_server` process
under control of a custom supervisor. This process responds
to the requestors itself, so the response time do not depend
on how busy ddoc_opener's message queue is.

Opener process for none revisioned ddocs made long-living. It does
a periodic check for a newest revision of its ddoc and in case
of an updated revision fetches the ddoc, keeping the cache warm.

BugsID: 28132

eiri committed:
Read LRU options from the couch config

LRU options now read from the couch config and not app's env.

The options could be specified in default.oni as the following:

    [ddoc_cache]
    max_objects = 50000
    max_size = 104857600
    max_lifetime = 60000

The defaults are 100 Mb for the max size of a cache with
no limitation on a number or max life-time of the documents.

BugzID: 28132

iilyak commented on src/ddoc_cache_fetcher_sup.erl:
```diff
@@ -0,0 +1,196 @@
+    catch
+        error:badarg ->
+            {error, {?MODULE, not_running}}
+    end.
```
The body of this function is almost the same as in `start_child`. I would implement `start_child` in terms of get_child. I.e. something like:

``` erlang
start_child(ChildId) ->
    case get_child(ChildId) of
       not_found -> gen_server:call(?MODULE, {start_child, ChildId}, ?TIMEOUT);
       Else -> Else
    end.
```

eiri replied:
Yes, sounds reasonable. I've added `get_child` on a second iteration, haven't thought to refactoring `start_child` then.

eiri committed:
Refactor `start_child` of fetch supervisor, bump doc_cache_open's vsn

eiri committed:
Remove match scaffolding from ddoc_cache_fetcher_sup

davisp commented:
Going to continue my review tomorrow. Its complex enough I wan to sit down and give it a thorough read.

eiri committed:
Address the comments in the original PR

eiri committed:
Make fetchers' sup more generic to allow it to manage different gen servers.

Refactor ddoc_cache_fetcher, move non-rev docs syncronization into its own gen_server process.

Catch the requests arrived after fetcher's child pulled the ddoc and
finished, but before the new ddoc inserted into the cache.

Catch the requests sent after fetcher already finished run,
but before its pid removed from the supervisor.

Save non-rev ddocs in the cache to do a lookup on them instead
of a cache traversal for the laterst revision.

Move specs defenitions into a common include file.

eiri committed:
Make the opener into a facade for the ddoc cache

Change all the ddoc interactions in fetcher and syncronizer to
work through the opener interface.

eiri committed:
Move from using ets_lru to custom ets cache implementation
